mrp2-utils/src/u256.rs

```diff
@@ -0,0 +1,949 @@
+            .fold(_false, |is_eq, (left_limb, right_limb)| {
+                let is_limb_equal = self.is_equal(left_limb.0, right_limb.0);
+                self.or(is_eq, is_limb_equal)
+            })
+    }
+
+    fn is_zero(&mut self, target: &UInt256Target) -> BoolTarget {
+        // since each limb is a 32-bit number, we can sum up the limbs without overflowing the native field.
+        // Therefore, for efficiency we determine whether `target` is zero by summing up its limbs and
+        // checking whether the sum is zero or not
+        let zero = self.zero();
+        let limbs_sum = target
+            .0
+            .iter()
+            .fold(zero, |sum, limb| self.add(sum, limb.0));
+        self.is_equal(limbs_sum, zero)
+    }
+
+    fn is_less_than_u256(&mut self, left: &UInt256Target, right: &UInt256Target) -> BoolTarget {
+        // left < right iff left - right requires a borrow
+        let (_, borrow) = self.sub_u256(left, right);
+        BoolTarget::new_unsafe(borrow.0)
+    }
+}
+
+impl<T: WitnessWrite<F>, F: RichField> WitnessWriteU256<F> for T {
+    fn set_u256_target(&mut self, target: &UInt256Target, value: U256) {
+        let mut bytes = [0u8; 32];
+        value.to_little_endian(&mut bytes);
+        let limbs = convert_u8_to_u32_slice(&bytes);
+        assert_eq!(limbs.len(), NUM_LIMBS);
+        target
+            .0
+            .iter()
+            .zip(limbs.iter())
+            .for_each(|(t, v)| self.set_target(t.0, F::from_canonical_u32(*v)));
+    }
+}
+
+impl<T: WitnessU32<F>, F: RichField> WitnessReadU256<F> for T {
+    fn get_u256_target(&self, target: &UInt256Target) -> U256 {
+        let bytes = target
+            .0
+            .iter()
+            .flat_map(|t| {
+                let (low, high) = self.get_u32_target(*t);
+                assert_eq!(high, 0); // check it is a 32-bit limb
+                low.to_le_bytes().to_vec()
+            })
+            .collect_vec();
+        U256::from_little_endian(&bytes)
+    }
+}
+
+impl UInt256Target {
+    /// Build a new `UInt256Target` from its limbs, provided in little-endian order
+    pub fn new_from_limbs(limbs: &[U32Target]) -> Result<Self> {
+        Ok(UInt256Target(limbs.try_into()?))
+    }
+    /// Utility function for serialization of UInt256Target
+    fn write_to_bytes(&self, buffer: &mut Vec<u8>) {
+        for i in 0..NUM_LIMBS {
+            buffer
+                .write_target(self.0[i].0)
+                .expect("Writing to a byte-vector cannot fail.");
+        }
+    }
+    /// Utility function for deserialization of UInt256Target
+    fn read_from_buffer(buffer: &mut Buffer) -> IoResult<Self> {
+        Ok(UInt256Target(
+            (0..NUM_LIMBS)
+                .map(|_| buffer.read_target().map(|t| U32Target(t)))
+                .collect::<Result<Vec<_>, _>>()?
+                .try_into()
+                .unwrap(),
+        ))
+    }
+}
+
+impl<'a> Into<Vec<Target>> for &'a UInt256Target {
```

Comment:
I found this function of little use, as one need to do complicated torsion like `<&Uint256Target as Into<Vec<Target>>>::into()` to be able to use it (at least Rust was not inferring automatically things for me when i tried to use it).
no need to change it now, but down the line, much better to simply use `impl From<UInt256Target> for Vec<Target>` (and same with reference) as we can directly call into() without complicated line.